benchmarks/bayesmark/optuna_optimizer.py

```diff
@@ -0,0 +1,116 @@
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Union
+
+import numpy as np
+
+from bayesmark.abstract_optimizer import AbstractOptimizer
+from bayesmark.experiment import experiment_main
+import optuna
+from optuna import pruners
+from optuna import samplers
+
+
+_SAMPLERS = {
+    "GridSampler": samplers.GridSampler,
+    "RandomSampler": samplers.RandomSampler,
+    "TPESampler": samplers.TPESampler,
+    "CmaEsSampler": samplers.CmaEsSampler,
+    "NSGAIISampler": samplers.NSGAIISampler,
+}
+
+_PRUNERS = {
+    "NopPruner": pruners.NopPruner,
+    "MedianPruner": pruners.MedianPruner,
+    "PatientPruner": pruners.PatientPruner,
+    "PercentilePruner": pruners.PercentilePruner,
+    "SuccessiveHalvingPruner": pruners.SuccessiveHalvingPruner,
+    "HyperbandPruner": pruners.HyperbandPruner,
+    "ThresholdPruner": pruners.ThresholdPruner,
+}
+
+
+Suggestion = Dict[str, Union[int, float]]
+ApiConfig = Dict[str, Dict[str, str]]
+
+
+class OptunaOptimizer(AbstractOptimizer):
+
+    primary_import = "optuna"
+
+    def __init__(self, api_config: ApiConfig, **kwargs: Any) -> None:
+
+        super().__init__(api_config, **kwargs)
+
+        try:
+            sampler = _SAMPLERS[kwargs["sampler"]]
+            sampler_kwargs: Dict[str, Any] = kwargs["sampler_kwargs"]
+
+        except KeyError:
+            raise ValueError("Unknown sampler passed to Optuna optimizer.")
+
+        try:
+            pruner = _PRUNERS[kwargs["pruner"]]
+            pruner_kwargs: Dict[str, Any] = kwargs["pruner_kwargs"]
+
+        except KeyError:
+            raise ValueError("Unknown pruner passed to Optuna optimizer.")
+
+        # We are using negative log-likelihood for classification
+        # and MSE for regression. Both are minimized.
+        self.study = optuna.create_study(
+            direction="minimize",
+            sampler=sampler(**sampler_kwargs),
+            pruner=pruner(**pruner_kwargs),
+        )
+        self.current_trials: Dict[int, int] = dict()
+
+    def _suggest(self, trial: optuna.trial.Trial) -> Suggestion:
+
+        suggestions: Suggestion = dict()
+        for name, config in self.api_config.items():
+            low, high = config["range"]
+            log = config["space"] == "log"
+
+            if config["space"] == "logit":
+                low = np.log(low / (1 - low))
+                high = np.log(high / (1 - high))
+
```

Comment:
This is not a strong opinion, but how about validating the config here? E,g, if `config["space"] == "logit"`, we need `0 < low <= high < 1`.

Reply:
Done in 569e1f1b51f5da664d4dde142b765e7df9c32abe.